src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -1859,23 +1870,10 @@ public override void List(CriteriaImpl criteria, IList results)
 
 				string[] implementors = Factory.GetImplementors(criteria.EntityOrClassName);
 				int size = implementors.Length;
-
```

Comment:
It looks to me there was already two loops with previous implementation, but one (the `UnionWith`) buried inside the other (the `for`).
I find this new implementation cleaner than previous one which was using a `for` loop. Now it causes more deviation from java impl and could cause harder port of Hibernate changes: is-it a trouble?